nupic/research/frameworks/vernon/mixins/centroid_context.py

```diff
@@ -45,22 +68,59 @@ def setup_experiment(self, config):
         dim_context = model_args.get("dim_context")
         input_size = model_args.get("input_size")
 
-        assert dim_context == input_size, ("For centroid experiments `dim_context` "
-                                           "must match `input_size`")
-
         super().setup_experiment(config)
 
-        # Tensor for accumulating each task's centroid vector
-        self.contexts = torch.zeros((0, self.model.input_size))
-        self.contexts = self.contexts.to(self.device)
+        centroid_context_args = config.get("centroid_context_args", {})
+        self.infer_while_training = centroid_context_args.get("infer_while_training",
+                                                              False)
+        self.subindices = None
+
+        if self.infer_while_training:
+
+            # Track clusters of data samples believed to be form the same task
+            self.clusters = []
+
+            # `contexts` is a list containing a single tensor since we require a
+            # mutable object
+            self.contexts = [torch.zeros((0, dim_context))]
+            self.n_samples_per_context = []
+
+            # Subsample features from input to form the context; we do this
+            # dimensionality reduction because in order to transform a T-squared
+            # variable to an F-distributed variable, the dimensionality of the context
+            # must be strictly less than the total number of data samples
```

Comment:
could that also explain why you don't necessarily have the 'perfect' number of clusters ?